runtime/userspace/api/caliptra-api/src/firmware_update/mod.rs

```diff
@@ -0,0 +1,294 @@
+// Licensed under the Apache-2.0 license
+
+extern crate alloc;
+mod pldm_client;
+mod pldm_context;
+mod pldm_fdops;
+
+use alloc::boxed::Box;
+use async_trait::async_trait;
+use caliptra_api::mailbox::{
+    CommandId, FwInfoResp, GetImageInfoResp, MailboxReqHeader, MailboxRespHeader,
+};
+use embassy_executor::Spawner;
+use flash_image::{FlashHeader, ImageHeader, CALIPTRA_FMC_RT_IDENTIFIER, SOC_MANIFEST_IDENTIFIER};
+use libsyscall_caliptra::mailbox::Mailbox;
+use libsyscall_caliptra::mailbox::{MailboxError, PayloadStream};
+use libtock_platform::ErrorCode;
+use libtockasync::TockExecutor;
+use pldm_common::message::firmware_update::get_fw_params::FirmwareParameters;
+use pldm_common::protocol::firmware_update::Descriptor;
+use pldm_lib::daemon::PldmService;
+use zerocopy::{FromBytes, Immutable, IntoBytes, KnownLayout};
+
+pub struct FirmwareUpdater<'a> {
+    staging_memory: &'static dyn StagingMemory,
+    mailbox: Mailbox,
+    params: &'a PldmFirmwareDeviceParams,
+    spawner: Spawner,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct PldmFirmwareDeviceParams {
+    pub descriptors: &'static [Descriptor],
+    pub fw_params: &'static FirmwareParameters,
+}
+
+impl<'a> FirmwareUpdater<'a> {
+    pub fn new(
+        staging_memory: &'static dyn StagingMemory,
+        params: &'a PldmFirmwareDeviceParams,
+        spawner: Spawner,
+    ) -> Self {
+        Self {
+            staging_memory,
+            mailbox: Mailbox::new(),
+            params,
+            spawner,
+        }
+    }
+
+    pub async fn start(&mut self) -> Result<(), ErrorCode> {
+        // Download firmware image to staging memory
+        pldm_client::initialize_pldm(
+            self.spawner,
+            self.params.descriptors,
+            self.params.fw_params,
+            self.staging_memory,
+        )
+        .await?;
+        pldm_client::pldm_wait_completion().await?;
+
+        // Parse the downloaded firmware image
+        let mut flash_header = [0u8; core::mem::size_of::<FlashHeader>()];
+        self.staging_memory
+            .read(0, &mut flash_header)
+            .await
+            .map_err(|_| ErrorCode::Fail)?;
+        let (flash_header, _) =
+            FlashHeader::read_from_prefix(&flash_header).map_err(|_| ErrorCode::Fail)?;
+        flash_header.verify().then_some(()).ok_or(ErrorCode::Fail)?;
+        let image_headers_offset = flash_header.image_headers_offset as usize;
+
+        // Update Caliptra
+        let (image_offset, image_len) = self
+            .get_image_toc(
+                flash_header.image_count as usize,
+                image_headers_offset,
+                CALIPTRA_FMC_RT_IDENTIFIER,
+            )
+            .await
+            .map_err(|_| ErrorCode::Fail)?;
+        self.update_caliptra(image_offset, image_len).await?;
+        self.wait_caliptra_rt_execution().await?;
```

Comment:
Maybe not in this PR, but we'll want to probably restart the user apps when updating, since Caliptra is going to reboot.

Caliptra updates mean that certs, cryptographic mailbox keys, DOT keys, etc. will be invalid and need to be recreated.

cc @parvathib